providers/ovh/ovhProvider.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/miekg/dns/dnsutil"
+	"github.com/xlucas/go-ovh/ovh"
+	"sort"
+	"strings"
+)
+
+type ovhProvider struct {
+	client *ovh.Client
+	zones  map[string]bool
+}
+
+var docNotes = providers.DocumentationNotes{
+	providers.DocDualHost:            providers.Can(),
+	providers.DocCreateDomains:       providers.Cannot("New domains require registration"),
+	providers.DocOfficiallySupported: providers.Cannot(),
+	providers.CanUseAlias:            providers.Cannot(),
+	providers.CanUseTLSA:             providers.Can(),
+	providers.CanUseCAA:              providers.Cannot(),
+	providers.CanUsePTR:              providers.Cannot(),
+}
+
+func newOVH(m map[string]string, metadata json.RawMessage) (*ovhProvider, error) {
+	appKey, appSecretKey, consumerKey := m["app-key"], m["app-secret-key"], m["consumer-key"]
+
+	c := ovh.NewClient(ovh.ENDPOINT_EU_OVHCOM, appKey, appSecretKey, consumerKey, false)
+
+	// Check for time lag
+	if err := c.PollTimeshift(); err != nil {
+		return nil, err
+	}
+
+	return &ovhProvider{client: c}, nil
+}
+
+func newDsp(conf map[string]string, metadata json.RawMessage) (providers.DNSServiceProvider, error) {
+	return newOVH(conf, metadata)
+}
+
+func newReg(conf map[string]string) (providers.Registrar, error) {
+	return newOVH(conf, nil)
+}
+
+func init() {
+	providers.RegisterRegistrarType("OVH", newReg)
+	providers.RegisterDomainServiceProviderType("OVH", newDsp, providers.CanUseSRV, providers.CanUseTLSA, docNotes)
+}
+
+func (c *ovhProvider) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	if err := c.fetchZones(); err != nil {
+		return nil, err
+	}
+	_, ok := c.zones[domain]
+	if !ok {
+		return nil, fmt.Errorf("%s not listed in zones for ovh account", domain)
+	}
+
+	ns, err := c.fetchNS(domain)
+	if err != nil {
+		return nil, err
+	}
+
+	return models.StringsToNameservers(ns), nil
+}
+
+type errNoExist struct {
+	domain string
+}
+
+func (e errNoExist) Error() string {
+	return fmt.Sprintf("Domain %s not found in your ovh account", e.domain)
+}
+
+func (c *ovhProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	dc.Punycode()
+	dc.CombineMXs()
+
+	if !c.zones[dc.Name] {
+		return nil, errNoExist{dc.Name}
+	}
+
+	records, err := c.fetchRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	var actual []*models.RecordConfig
+	for _, r := range records {
+		if r.FieldType == "SOA" {
+			continue
+		}
+
+		if r.SubDomain == "" {
+			r.SubDomain = "@"
+		}
+
+		// ovh uses a custom type for SPF and DKIM
+		if r.FieldType == "SPF" || r.FieldType == "DKIM" {
+			r.FieldType = "TXT"
+		}
+
+		// ovh default is 3600
+		if r.TTL == 0 {
+			r.TTL = 3600
+		}
+
+		rec := &models.RecordConfig{
+			NameFQDN:       dnsutil.AddOrigin(r.SubDomain, dc.Name),
+			Name:           r.SubDomain,
+			Type:           r.FieldType,
+			Target:         r.Target,
+			TTL:            uint32(r.TTL),
+			CombinedTarget: true,
+			Original:       r,
+		}
+		actual = append(actual, rec)
+	}
+
```

Comment:
Insert this (all providers now must do this; though if you were passing the integration tests it might be a no-op):

```
  // Normalize
  models.Downcase(actual)
```

Reply:
Oh, I missed this change, I'll do that.